pkg/providers/instance/resperrorhandlers_test.go

```diff
@@ -35,15 +39,112 @@ import (
+	errMsgAllocationFailureFmt         = "unable to allocate resources with selected VM size (%s). (will try a different VM size to fulfill your request)"
+	errMsgOverconstrainedZonalFmt      = "unable to allocate resources in the selected zone (%s) with %s capacity type and %s VM size. (will try a different zone, capacity type or VM size to fulfill your request)"
+	errMsgOverconstrainedAllocationFmt = "unable to allocate resources in all zones with %s capacity type and %s VM size. (will try a different capacity type or VM size to fulfill your request)"
+	errMsgRegionalQuotaExceeded        = "regional on-demand vCPU quota limit for subscription has been reached. To scale beyond this limit, please review the quota increase process here: https://learn.microsoft.com/en-us/azure/quotas/regional-quota-requests"
+)
+
+// offering represents a zone and capacity type combination for cleaner test setup
+type offering struct {
+	zone         string
+	capacityType string
+}
+
+// Common offering configurations
+var (
+	zone1OnDemand = offering{zone: testZone1, capacityType: karpv1.CapacityTypeOnDemand}
+	zone1Spot     = offering{zone: testZone1, capacityType: karpv1.CapacityTypeSpot}
+	zone2OnDemand = offering{zone: testZone2, capacityType: karpv1.CapacityTypeOnDemand}
+	zone2Spot     = offering{zone: testZone2, capacityType: karpv1.CapacityTypeSpot}
+	zone3OnDemand = offering{zone: testZone3, capacityType: karpv1.CapacityTypeOnDemand}
+	zone3Spot     = offering{zone: testZone3, capacityType: karpv1.CapacityTypeSpot}
 )
 
+type testCaseBuilder struct {
+	tc responseErrorTestCase
+}
+
+func newTestCase(name string) *testCaseBuilder {
+	return &testCaseBuilder{
+		tc: responseErrorTestCase{
+			testName:                                name,
+			originalRequestSKU:                      createDefaultTestSKU(),
+			expectedUnavailableOfferingsInformation: []offeringToCheck{},
+			expectedAvailableOfferingsInformation:   []offeringToCheck{},
+		},
+	}
+}
+
+func (b *testCaseBuilder) withInstanceType(offerings ...offering) *testCaseBuilder {
+	offeringTypes := make([]struct{ zone, capacityType string }, len(offerings))
+	for i, o := range offerings {
+		offeringTypes[i] = struct{ zone, capacityType string }{zone: o.zone, capacityType: o.capacityType}
+	}
+	b.tc.instanceType = defaultCreateInstanceType(offeringTypes...)
+	return b
+}
+
+func (b *testCaseBuilder) withEmptyInstanceType() *testCaseBuilder {
+	b.tc.instanceType = &cloudprovider.InstanceType{}
+	return b
+}
+
+func (b *testCaseBuilder) withZoneAndCapacity(zone, capacityType string) *testCaseBuilder {
+	b.tc.zone = zone
+	b.tc.capacityType = capacityType
+	return b
+}
+
+func (b *testCaseBuilder) withResponseError(errorCode, errorMessage string) *testCaseBuilder {
+	b.tc.responseErr = createResponseError(errorCode, errorMessage)
+	return b
+}
+
+func (b *testCaseBuilder) withNilResponseError() *testCaseBuilder {
```

Comment:
Seems implied/default?

Any actual need for it?

Reply:
Yeah can be removed, good point!